orangecontrib/timeseries/widgets/owdifference.py

```diff
@@ -30,186 +37,152 @@ class Inputs:
+            controlWidth=60, alignment=Qt.AlignRight,
+            callback=self.commit.deferred,
+            tooltip="Sets the distance between points; 1 for consecutive points"
+        )
+        gui.rubber(sp.box)
+
+        gui.checkBox(
+            box, self, 'invert_direction', 'Invert differencing direction',
+            callback=self.commit.deferred)
+
+        gui.auto_commit(self.buttonsArea, self, 'autocommit', '&Apply')
+
+    def _operation_changed(self):
+        self.controls.shift_period.box.setEnabled(self.operation != self.Diff2)
+        self.commit.deferred()
+
+    def _selection_changed(self):
+        self.selection = [
+            self.model.data(index)
+            for index in self.view.selectionModel().selectedRows()]
+        self.commit.deferred()
 
     @Inputs.time_series
     def set_data(self, data):
-        self.closeContext()
-        self.data = data = None if data is None else \
-                           Timeseries.from_data_table(data)
-        if data is not None:
-            self.model[:] = [var for var in data.domain.variables
-                             if var.is_continuous and var is not
-                             data.time_variable]
-            self.select_default_variable()
-            self.openContext(self.data)
-            self._restore_selection()
-        else:
-            self.reset_model()
-        self.on_changed()
-
-    def _restore_selection(self):
-        def restore(view, selection):
-            with signal_blocking(view.selectionModel()):
-                # gymnastics for transforming variable names back to indices
-                var_list = [var for var in self.data.domain.variables
-                            if var.is_continuous and var is not
-                            self.data.time_variable]
-                indices = [var_list.index(i) for i in selection]
-                select_rows(view, indices)
-        restore(self.view, self.selected)
-
-    def select_default_variable(self):
-        self.selected = [0]
-        select_rows(self.view, self.selected)
-
-    def reset_model(self):
-        self.model.wrap([])
-
-    def on_changed(self):
-        var_names = [i.row()
-                     for i in self.view.selectionModel().selectedRows()]
-        self.order_spin.setEnabled(
-            self.shift_period == 1
-            and self.chosen_operation == self.Operation.DIFF)
-        self.selected = [self.model[v] for v in var_names]
-        self.commit()
+        if self.selection:
+            self.persistent_selection = self.selection[:]
+
+        if not data:
+            self.data = None
+            self.model.clear()
+            self.commit.now()
+            return
+
+        self.data = Timeseries.from_data_table(data)
+        self.model[:] = [var for var in data.domain.variables
+                         if var.is_continuous and var is not
+                         data.time_variable]
```

Comment:
I see the culprit. Before, `self.data` was also assigned to `data` variable, which is already the converted table. You convert the table above, but you still call `data.time_variable`, which doesn't necessarily exist. I think calling `self.data.time_variable` should solve the issue.

Reply:
Thanks! I also added a test, and preview now sends Iris as-is, without converting it to time series (so widget has to do it).